### ISSUE TITLE
[Tracking] Re-rewrite Executer

### ISSUE THREAD

**Issue by astronouth7303** [xonsh/xonsh#2159]
Xonsh is having a lot of trouble with fundamental low-level stuff (unthreadable, misalignment of standard out/error, signal issues, etc). Let's fix it.

Some of the issues come from using `subprocess` as the underlying executer, and others from conflating interfacing with the OS with the process of turning an abstract pipeline into concrete system objects.

1. [ ] Write a abstraction library to deal with OS API issues and generally manage resources. ([xonsh/slug](https://github.com/xonsh/slug))
2. [ ] Rewrite xonsh to vendor and use said abstraction library.

**Comment by mitnk** [comment]
IMHO, using `subprocess` may not be a bad thing. After all it's tweaked in years by the community for various platforms.

However `xonsh/slug` looks cool!

**Comment by astronouth7303** [comment]
It crossed my mind to use `subprocess.Popen` as the underlying implementation of a Process primitive. But we still need good management of process groups, integration of virtual processes (Python code as a "process"), and some well-tested utility pipes (tapped pipes, going between pipes and buffers, etc).

**Comment by mitnk** [comment]
I'd say this `xonsh/slug` may go through all the hard things which the current or the previous job controls / subproc running. But you can try this direction out, though it may as hard as the current one.

Another obvious direction is fix issues based on what systems we currently have, it works at over 90% cases, and @scopatz and users have spent countless time to implementing & testing it.

> Rewrite xonsh to vend and use said abstraction library.

For this part, I'd like to know @scopatz 's idea: do you think it's ok to make xonsh's exec-or job-control part replaceable, like history backends :) Anyway this is the heart of xonsh (except parsing part).

**Comment by scopatz** [comment]
Thanks for the concern @mitnk. For the record I am all for any efforts to make the code base more modular and maintainable. The sorts of things that @astronouth7303 is proposing have a lot of merit. 

My understanding / guess is that the initial version of slug would essentially be a copy of what we have in xonsh right now anyway. Also, now that we have an executor that *mostly* works, I think it is smart to do further development outside of mainline xonsh. That way we won't block releases for a couple of months or risk breaking everything until we really know it works. 

> xonsh's exec-or job-control part replaceable, like history backends

Well I don't think it will be replaceable like the history backends or shell frontends. It is going to be more like lazyasd, where there is a separate project that is a core component of xonsh that other people can use if they want. We'll still only have one execution model. At least that was how I understood the plan.

**Comment by astronouth7303** [comment]
Pretty much exactly, yeah. I am planning on building bullet-proof primitives and then refactor xonsh to make use of them. @scopatz and I discussed this and modularization in general before hand; I didn't just wake up and decide to do this.

**Comment by mitnk** [comment]
Hah, That's a relief :D

**Comment by laerus** [comment]
It's worth taking a look at the `multiprocessing` library, it doesn't seem to use `subprocess` and uses threads for IPC too, we can borrow some ideas for aliases and whatnot. 

for me the ideal would be to use as little threaded code as possible,
 on UNIX `select` based subprocess io feels nicer,
(maybe even xonsh gets a native event loop)

**Comment by astronouth7303** [comment]
@laerus I think the direction we're going is native unmanaged pipes. Ideally, any threads that are necessary should be managed and transparent. All xonsh should need to do is just wait on the process group.

`multiprocessing` has some use, but it mostly just `fork()`s all over. The trick is that to run a program under posix:

1. `fork()`
2. In some order:
   * Set up the environment variables
   * Copy the appropriate file descriptors to standard in/out/error
   * Close the file descriptors that aren't marked non-inheritable (I suspect this varies widely by platform)
   * Possibly do other set up (process groups, control groups, creating a session, etc)
3. `exec()` the program you're actually trying to run.

There's a reason I'd like to use at least some `subprocess`.

**Comment by scopatz** [comment]
@laerus, as @astronouth7303 mentioned, it is basically a fork(). This can use up a lot of system resources in Python due to the copy-on-write behaviour.  

Importantly, though, the fork idea was the first idea that we tried and then it didn't work. This was the underlying implementation of the now-gone TeePTY class.  While this sort of worked for its intended purpose of capturing and teeing output, it didn't work on Windows and failed miserably for job control. I think this is because signals had be passed from the main Python process (which received the signal), to the forked Python process, and then finally down to the subprocess we are running. Given that fork literally copies the process memory, we never got it so that the signal handlers were replaced correctly in both interpreters. Additionally confounding that is that Python makes you handle signals on the main thread (even though there is a GIL, which I never understood).

So in summary, I think going with a fork-based solution is probably not correct because it is so resource intensive and we couldn't get it working when we tried. As always I am open to being proven wrong :)

**Comment by astronouth7303** [comment]
To be clear, you still need to `fork()` on POSIX. [subprocess](https://github.com/python/cpython/blob/9574e0adf287e11ffd828e60c984284339d4faa2/Modules/_posixsubprocess.c#L672) does this on POSIX as well. But POSIX has always been like this and has copy-on-write semantics for forked processes, making the fork itself very cheap. (There is some expense as data pages are written to, but it's much less than copying potentially gigabytes of pages. For many fork-exec situations, only the stack would need to be copied. Having to drop back into Python kinda wrecks this, but it's still probably cheaper.)

But on win32, there's no `exec()` or `fork()`. (Don't ask me how cygwin etc do it). Instead, there's [much richer native APIs](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425(v=vs.85).aspx) that pretty much handle our needs.

**Comment by astronouth7303** [comment]
Ok, slug is either at or very close to MVP, so it's about time to start thinking about how xonsh is going to use this.

Only a few things are missing:

* Virtual Processes
* Psuedo Terminals

Both of these are fairly problematic, so I'm going to have to spend some time looking into them. (Eg, how do you terminate/kill a Python thread? Do ptys work on mac? etc)

**Comment by laerus** [comment]
i think RHettinger talks about not killing a thread and why they didnt provide such functionality in his talk 'thinking about concurrency'

**Comment by astronouth7303** [comment]
@laerus I'll have to look into it again. Unfortunately, it's either figure out a way to get a thread to exit, make aliases into coroutines, or suck it up and hope all aliases exit when their pipes quit on them.

**Comment by astronouth7303** [comment]
Oh, nvm on option 1. Python's `.read()` can't be cancelled externally except on my the main thread by SIGINT and similar.

**Comment by whitelynx** [comment]
Is anything happening with this? It looks like `slug` hasn't been touched in 9 months.

**Comment by scopatz** [comment]
Yeah, I thinbk slug development has stalled a bit. There are some really tricky cases to cover.

**Comment by anki-code** [comment]
@AstraLuma May be there are easier ways?

* Reduce cross platform
* Reduce Python (use prebuilt helpers or bash)
* Make fork from Python and fix subprocess
* Using OS/kernel features... I don't know

I mean that what if try to think about crazy/bad but guarantee stable ways.

**Comment by AstraLuma** [comment]
1. it's one of the big selling points
2. Good Luck With That
3. I actually looked into this when I first wrote slug.
4. You should look at the slug implementation

Honestly, there's a few big problems:

1. Xonsh can't keep separating stdout and stderr
2. I just lost motivation trying to untangle the mess of code that is the executor, and trying to reimplantat all the pieces
3. There's restrictions in the core of python that prevent you killing other threads. I don't fully understand why that is or how you would change it. (I looked into it once upon a time.) Forking and fixing is just as much mucking around in the core internals, plus you now have a fork to maintain.

I have also since discovered edge casey bugs in slug, so that needs fixing.

**Comment by anki-code** [comment]
May be [the plumbum project](https://github.com/tomerfiliba/plumbum) have new ideas? Could anyone take a short look?

**Comment by anki-code** [comment]
@AstraLuma Just to let you know that if you have some thoughts but have no time I can sponsoring the work because believe that xonsh is great and the executor is the core part.

**Comment by anki-code** [comment]
@AstraLuma 

> There's restrictions in the core of python that prevent you killing other threads. I don't fully understand why that is or how you would change it.

I made steps in the direction of killing the threads. 

First of all I tried to use `slug` but without success because the example in the readme is not working in my `echo 1 | grep 1` example. After a bit review of the slug code I thought I should try from scratch to understand more context.

All the solutions that was [described on stackoverflow](https://stackoverflow.com/questions/4789837/how-to-terminate-a-python-subprocess-launched-with-shell-true/4791612#4791612) is not working for me on Linux and it's not regarding `shell=True` or not.

Finally I've tried to use `psutil` for this and have some expected success. The test is: create `big.txt` file with many lines and run the process `cat big.txt` and interrupt them during the printing lines. The success is interrupting the printing the file without waiting the process end.

Here is the PoC that works for me on Linux with Python 3.8 and psutil 5.8.0:

```python
xonsh
mkdir -p /tmp/prc
cd /tmp/prc

# Create big file
with open('big.txt', 'w') as f:
    for i in range(0, 1_000_000):
        print(i, file=f)

        
# From https://psutil.readthedocs.io/en/latest/index.html?highlight=kill#kill-process-tree
import os
import signal
import psutil

def kill_proc_tree(pid, sig=signal.SIGTERM, include_parent=True,
                   timeout=None, on_terminate=None):
    """Kill a process tree (including grandchildren) with signal
    "sig" and return a (gone, still_alive) tuple.
    "on_terminate", if specified, is a callback function which is
    called as soon as a child terminates.
    """
    assert pid != os.getpid(), "won't kill myself"
    parent = psutil.Process(pid)
    children = parent.children(recursive=True)
    if include_parent:
        children.append(parent)
    for p in children:
        try:
            p.send_signal(sig)
        except psutil.NoSuchProcess:
            pass
    gone, alive = psutil.wait_procs(children, timeout=timeout,
                                    callback=on_terminate)
    return (gone, alive)

    
import os, subprocess, time
p = subprocess.Popen(['cat', 'big.txt'])
# or
# p = subprocess.Popen('cat big.txt', shell=True)  # we'll have two killed (shell + child)
    
time.sleep(0.01)    

print(kill_proc_tree(p.pid))
print('done')
```
The output:
```
0
1
2
3
4
...
...
...
47184
killed: ([psutil.Process(pid=402463, status='terminated', exitcode=<Negsignal.SIGTERM: -15>, started='11:35:47')], [])
done
```

Does it make sense?

PS: Please give me more description of the problems that I/We can try to solve.

**Comment by anki-code** [comment]
I see that `psutil` and `slug` use `os.kill()`. Agrh... Now I don't understand what this mean without repeatable example:

> There's restrictions in the core of python that prevent you killing other threads.

If `slug` works for Linux on my example above, what is the problem?

**Comment by AstraLuma** [comment]
Killing other processes is fine and great.

Problems and complexities come from what slug calls virtual processes--callable aliases in xonsh.

**Comment by anki-code** [comment]
I you mean that we have processes for the tools in pipe but not have the process for callable alias... what if we will do replace `lambda args: print(1)` to smth like `xonsh -c print(1)` while creating the pipeline? Speed issues?

**Comment by anki-code** [comment]
@AstraLuma why this answers is not working - https://stackoverflow.com/a/325528/14470020 ?

**Comment by AstraLuma** [comment]
It's late and my brain demands I get this out.

I consider the current situation of xonsh re threadable & unthreadable processes. It's bad UX and no other shell has this trade-off. The constant stream of bugs and support requests around this existed when i joined the project in 2016. It's been a thorn in my side as recently as this month.

if a process is threadable, you can use it in pipelines and do code with it, but it might have problems if used interactively. If it is unthreadable, it behaves perfectly for live user interaction, but it couldn't be used for scripting. And software like ssh and git exist which can be (and often are) used both ways.

I decided to fix this by greenfielding the core of how xonsh actually executes processes, and started work on Slug.

Slug (a snail without a shell, because xonsh imagry) was supposed to be a new core for xonsh. A bulletproof and independently tested (and relatively easy to test) library providing the core primitives around processes, pipelines, process groups--the stuff shells are most known for managing. (It also included a concept called Virtual Process: a duck-typed Process that didn't intrinsically create resources, just let the application inject code into the process framework.) If we ever had bugs around signals, piping, or anything else, we could reproduce it as slug calls and turn it into a test.

Oh, and be cross platform as well. (My favorite is still learning that windows can natively pause and continue processes.)

I got slug to the point that it worked. Not perfectly; I would discover bugs later. But it worked well enough that I got started on the xonsh side of the work.

I don't remember all the specifics--4 years is a long time for me--but I got the jist of it working and then got bogged down in some bugs and some weird behaviors. I don't remember if some of the advanced pipeline tools in slug didn't behave right, or if virtual processes caused more problems than I realized, or if there some hard-to-track down oddity with signals. (I do know that I tried to handle terminal signals in virtual processes by injecting an event into the thread--there's a C API for that--but for some reason it got swallowed or something. I learned in the last few months that [dramatiq](https://dramatiq.io/) pulls a similar trick, but I haven't revisited it because honestly? That was not the biggest problem.)

The way this effort dragged on--and pretty much having to do it alone--is the big reason I drifted from the xonsh project and am no longer active.

**Comment by anki-code** [comment]
> and pretty much having to do it alone

I believe that in case of complex work the success and the community help could be achieved by doing this:

1. The clean goal that clearly described.

2. The step by step guide how we will achieve the goal as far as we see. What we will do and what we will not do on every step: minimum viable product on every step. 

3. Repeatable pieces of code and repeatable pieces of bugs clearly described. Anybody from the street can grab the code, then repeat the bug by copying and pasting the commands and then dive into it. This is 10% to repeat and 90% to understand. 

4. If we have no points 1-2 then we should have point 3 as the minimum.

In the issues around xonsh I see some unfocused buzz between a few genius people without some code, clean description of bugs and links. It's very hard to dive into this from the street and start helping. Very high entry threshold. 

<details>

![image](https://user-images.githubusercontent.com/1708680/115986735-572d0380-a5ba-11eb-80d9-38d113a79057.png)

</details>

**Comment by AstraLuma** [comment]
I'm not sure how much the work can be broken up and how much just needs to be done in one herculean effort.

A lot of the features of xonsh are built on top of the executor. You can't save implementing those features until after some MVP.

Like, in #2337, I touched ~3000 lines of code and I don't remember anything that could be punted or saved for a secondary PR. I think that was all core stuff.

**Comment by anki-code** [comment]
While I was in the issues around 4214 I made some big picture:

![image](https://user-images.githubusercontent.com/1708680/116824270-fbeaaa80-ab91-11eb-8ab0-378c3ed93507.png)

**Comment by anki-code** [comment]
> A lot of the features of xonsh are built on top of the executor. You can't save implementing those features until after some MVP.

I think the good way is to start from the list of supported use cases. For example:
1. `echo 1` - single command
2. `echo 1 | grep 1` - pipe
3. `echo 1 &` - daemon
4. `echo $(fzf)` - interactive capturable
5. etc

Then answer: what example the slug can cover?
Then isolate the example and implement the slug only for this example in xonsh and make it optional.
